File: zig/413.zig
const std = @import("std");

fn solution(nums: []const i32) u32 {
    var result: u32 = 0;

    var i: usize = 2;

    while (i < nums.len) : (i += 1) {
        const diff = nums[i] - nums[i - 1];

        if (nums[i - 1] - nums[i - 2] == diff) {
            var ct: u32 = 1;
            var last = nums[i];

            var j = i + 1;

            while (j < nums.len) : (j += 1) {
                if (nums[j] - last != diff) {
                    break;
                }

                ct += 1;
                last = nums[j];
            }

            i = j - 1;
            result += (ct * (ct + 1)) / 2;
        }
    }

    return result;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const input = [_]i32{ 1, 2, 3, 4 };
    const expected = 3;

    try expect(solution(&input), expected);
}

test "test case #2" {
    const input = [_]i32{1};
    const expected = 0;

    try expect(solution(&input), expected);
}

File: go/413.go
package main

func numberOfArithmeticSlices(nums []int) int {
	result := 0

	var i = 2

	for ; i < len(nums); i++ {
		diff := nums[i] - nums[i-1]
		if nums[i-1]-nums[i-2] == diff {
			ct := 1
			last := nums[i]

			j := i + 1

			for ; j < len(nums); j++ {
				if nums[j]-last != diff {
					break
				}

				ct++
				last = nums[j]
			}

			i = j - 1
			result += (ct * (ct + 1)) / 2
		}

	}

	return result
}
